### ISSUE TITLE
Improved error message when trying to update static leaves

### ISSUE THREAD

**Issue by LouisDesdoigts** [patrick-kidger/equinox#452]
Hey Patrick, I've been playing around with the idea of enabling the updating of static leaves in my downstream library (details [here](https://github.com/LouisDesdoigts/zodiax/issues/29) if you want to look). 

Long story short making this safer requires catching the empty `AssertionError` that is thrown [here](https://github.com/patrick-kidger/equinox/blob/main/equinox/_tree.py#L25). I want to make this more robust by checking for the specific error message to ensure other errors aren't accidentally caught. Another possibility could be to raise a custom `StaticLeaf` error or something like that. 

Here a minimal example:
```python
import equinox as eqx

class Foo(eqx.Module):
    bar : str = eqx.field(static=True)

    def __init__(self):
        self.bar = 'baz'

eqx.tree_at(lambda tree: tree.bar, Foo(), 'new_baz')
```

which throws the empty `AssertionError`.

I was thinking something simple for the error message like "Static leaves can not be updated", but there might be some other cases that this is catching that I'm unaware of.

Anyway let me know what you think, I'm happy to do the PR for this.

Thanks as always for your help!

**Comment by patrick-kidger** [comment]
Yeah, using `tree_at` to update something that isn't even part of the pytree isn't defined. It's like trying to change a method on a dictionary: `eqx.tree_at(lambda d: d.get, dict(), None)`.

I agree we could make the error message clearer. I'd be happy to take a PR on this.

**Comment by patrick-kidger** [comment]
Fixed in #460.
